Fix issues regarding aosegs/aocssegs in state AOSEG_STATE_AWAITING_DROP

jimmyyih opened the pull request:
Commit message dump for context:
```
commit fb17a31406d89d82914774e6f3479e0dff7208e2
Author: Jimmy Yih <jyih@pivotal.io>
Date:   Tue May 2 14:53:57 2017 -0700

    VACUUM should cleanup aosegs/aocssegs in AOSEG_STATE_AWAITING_DROP state

    Due to the different phases of VACUUM, compaction phase could have
    committed but the drop was cancelled. This could leave some
    aoseg/aocsseg entries in AOSEG_STATE_AWAITING_DROP state which could
    then be cleaned up in a subsequent VACUUM as expected... but it
    currently just skips. The master keeps a hash table (AppendOnlyHash)
    and uses it during VACUUM to check tuple count, state, and
    aborted/committed. Our issue here is that state checking was only
    being done against state AWAITING_DROP_READY. It was not including
    COMPACTED_AWAITING_DROP in the state check. This commit adds the check
    and updates the detailed comment block on the state transitions in
    appendonlywriter.h.

commit c3c77bf5e71bab331d21ff18b7f3e995b18d8c00
Author: Jimmy Yih <jyih@pivotal.io>
Date:   Tue May 2 14:52:48 2017 -0700

    ADD COLUMN on an AOCO table should check and drop aocssegs in await drop

    When doing an ADD COLUMN operation on an AOCO table, the vpinfo byte
    array stored in the aocsseg auxiliary table for each column is updated
    to add an extra vpinfo index to match the relnatts value in
    pg_class. However, entries in the aocsseg auxiliary table with state 2
    (AOSEG_STATE_AWAITING_DROP) will not have their vpinfo byte array
    updated. Subsequent calls to getAOCSVPEntry function will then fail
    due to the vpinfo size does not match the relnatts value in pg_class
    (e.g. persistent table rebuild). Since we have the AccessExclusiveLock
    on the table during ADD COLUMN, we might as well schedule the drop
    before the vpinfos are updated to prevent any inconsistency issues.
```

jimmyyih commented on src/test/isolation2/expected/add_column_after_vacuum_cancel_column.out:
```diff
@@ -0,0 +1,61 @@
+2U: SELECT segno, column_num, state FROM gp_toolkit.__gp_aocsseg_name('ao');
+segno|column_num|state
+-----+----------+-----
+1    |0         |1    
```
That's what I thought too... but segno1 actually gets recreated after it is dropped.  This is the same in vacuum full too:
```
gptest=# create table foobar (a int, b int) with (appendonly=true, orientation=column);
NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
CREATE TABLE
gptest=# insert into foobar select i, i from generate_series(1,10)i;
INSERT 0 10
gptest=# delete from foobar;
DELETE 10
gptest=# select gp_toolkit.__gp_aocsseg_name('foobar') from gp_dist_random('gp_id');
        __gp_aocsseg_name
---------------------------------
 ("(0,2)",1,0,1,2,48,48,1,3,1)
 ("(0,2)",1,1,129,2,48,48,1,3,1)
 ("(0,2)",1,0,1,5,64,64,1,3,1)
 ("(0,2)",1,1,129,5,64,64,1,3,1)
 ("(0,2)",1,0,1,3,56,56,1,3,1)
 ("(0,2)",1,1,129,3,56,56,1,3,1)
(6 rows)

gptest=# vacuum full foobar;
VACUUM
gptest=# select gp_toolkit.__gp_aocsseg_name('foobar') from gp_dist_random('gp_id');
       __gp_aocsseg_name
-------------------------------
 ("(0,1)",1,0,1,0,0,0,1,3,1)
 ("(0,1)",1,1,129,0,0,0,1,3,1)
 ("(0,3)",2,0,2,0,0,0,1,3,1)
 ("(0,3)",2,1,130,0,0,0,1,3,1)
 ("(0,1)",1,0,1,0,0,0,1,3,1)
 ("(0,1)",1,1,129,0,0,0,1,3,1)
 ("(0,3)",2,0,2,0,0,0,1,3,1)
 ("(0,3)",2,1,130,0,0,0,1,3,1)
 ("(0,1)",1,0,1,0,0,0,1,3,1)
 ("(0,1)",1,1,129,0,0,0,1,3,1)
 ("(0,3)",2,0,2,0,0,0,1,3,1)
 ("(0,3)",2,1,130,0,0,0,1,3,1)
(12 rows)
```

jimmyyih commented:
Adding another somewhat related fix to this PR:
```
commit cfc5b26843e782bf84a4f666b4a67658270db03a
Author: Jimmy Yih <jyih@pivotal.io>
Date:   Thu May 4 15:39:39 2017 -0700

    Fix lazy VACUUM to not wait for AccessExclusiveLock during drop phase

    We expect lazy VACUUM to skip drop phase if it cannot acquire the
    relation's AccessExclusiveLock. This commit properly sets the dontWait
    flag to True when calling try_relation_open during drop phase. We also
    change the dontWait part of try_relation_open to not error out because
    callers should be the ones to handle an invalid relation.
```

Refactored isolation2 tests from the first two bug fix commits since lazy VACUUM won't block now.  Those tests technically cover this lazy VACUUM fix... hence adding the fix to this PR (when merging into master, the lazy VACUUM fix will be rebased to go in first).

ashwinstar commented on src/test/isolation2/sql/add_column_after_vacuum_skip_drop_column.sql:
```diff
@@ -7,13 +7,13 @@ INSERT INTO ao SELECT i as a, i as b FROM generate_series(1, 10) AS i;
 1: SELECT COUNT(*) FROM ao;
-2U: SELECT segno, column_num, state FROM gp_toolkit.__gp_aocsseg_name('ao');
-2>: VACUUM ao;
-3: select kill_vacuum();
```
Please modify the `@Description` of the test on top.

ashwinstar commented on src/backend/commands/tablecmds.c:
```diff
@@ -4565,21 +4566,35 @@ ATAocsWriteNewColumns(
+		 * AOSEG_STATE_AWAITING_DROP. At the end of the loop, we will
+		 * try to drop the segs since we currently have the
+		 * AccessExclusiveLock. If we don't do this, an aocsseg table
+		 * could have issues due to the difference in vpinfo size and
```
Instead of saying could have issues, please can we clearly specify what problem happens like vpinfo size contains info for less number of columns compared to relnatts.

ashwinstar commented on src/backend/commands/tablecmds.c:
```diff
@@ -4593,6 +4608,10 @@ column_to_scan(AOCSFileSegInfo **segInfos, int nseg, int natts)
 	}
+
+	if (list_length(drop_segno_list) > 0 && Gp_role != GP_ROLE_DISPATCH)
+		AOCSDrop(aocsrel, drop_segno_list);
```
I am missing to understand the `AOCSDrop()` interface, like why does it need `drop_segno_list`. As its internally anyways scanning the full aocsseg table and checking for `AOSEG_STATE_AWAITING_DROP` state what real purpose the `drop_segno_list` is serving.

jimmyyih replied:
@asimrp: It should not touch regular segfiles in `AOCSDrop()` with this change.  Inside the `AOCSDrop()` loop, there is a check to see if the segfile is in our drop_segno_list where it will skip that iteration if not.  Only segfiles in state `AOSEG_STATE_AWAITING_DROP` that we found in `column_to_scan()` will be operated on in the `AOCSDrop()` loop.  It's a bit inefficient since we're basically looping the aocsseg table twice in thise code flow but I decided on using existing `AOCSDrop()` function instead of creating my own drop function.

@ashwinstar: `AOCSDrop()` is used primarily in `vacuum_appendonly_rel()` where the segno list is generated by the query dispatcher and sent to the query executors (QE) to do their drops or to truncate the segfile to EOF.  For this fix, I am reusing the `AOCSDrop()` function instead of creating a new drop function so each QE needs to generate their own `drop_segno_list` (which is correct as we have seen cases where only one primary/mirror segment pair had an aocssegfile in state `AOSEG_STATE_AWAITING_DROP` but other primary/mirror segments were fine).  So mainly the `drop_segno_list` is serving as required input to reuse the `AOCSDrop()` function.

ashwinstar commented on src/test/isolation2/expected/add_column_after_vacuum_skip_drop_column.out:
```diff
@@ -0,0 +1,57 @@
+-- @Description Ensures that an alter table add column after a canceled vacuum operation is ok
+--
+DROP TABLE IF EXISTS ao;
```
We don't need to do this, and use some non-generic name please.

ashwinstar commented on src/test/isolation2/expected/vacuum_after_vacuum_skip_drop_column.out:
```diff
@@ -0,0 +1,55 @@
+-- @Description Ensures that a vacuum after a canceled vacuum operation is ok
+--
+DROP TABLE IF EXISTS ao;
```
Same here

ashwinstar commented on src/test/isolation2/expected/add_column_after_vacuum_skip_drop_column.out:
```diff
@@ -1,18 +1,17 @@
--- @Description Ensures that an alter table add column after a canceled vacuum operation is ok
+-- @Description Ensures that an alter table add column after vacuum skips
```
Well the test is only ensuring "add column drops segfiles in AOSEG_STATE_AWAITING_DROP" and not really before or after updating segfile vpinfos.

ashwinstar commented on src/test/isolation2/expected/vacuum_after_vacuum_skip_drop_column.out:
```diff
@@ -1,18 +1,17 @@
--- @Description Ensures that a vacuum after a canceled vacuum operation is ok
+-- @Description Ensures that a vacuum after vacuum skips drop phase will
```
Instead "Ensures next vacuum drops segfiles in AOSEG_STATE_AWAITING_DROP state, left by previous vacuum in that state."

ashwinstar commented:
On Fri, May 5, 2017 at 2:43 PM, Jimmy Yih <notifications@github.com> wrote:
@ashwinstar <https://github.com/ashwinstar>: AOCSDrop() is used primarily
in vacuum_appendonly_rel() where the segno list is generated by the query
dispatcher and sent to the query executors (QE) to do their drops or to
truncate the segfile to EOF. For this fix, I am reusing the AOCSDrop()
function instead of creating a new drop function so each QE needs to
generate their own drop_segno_list (which is correct as we have seen cases
where only one primary/mirror segment pair had an aocssegfile in state
AOSEG_STATE_AWAITING_DROP but other primary/mirror segments were fine). So
mainly the drop_segno_list is serving as required input to reuse the
AOCSDrop() function.

Thanks, makes sense, please can you add comment on top of AOCSDrop() to
explain the same.

asimrp commented on src/backend/commands/test/tablecmds_test.c:
```diff
@@ -73,7 +73,7 @@ test__column_to_scan(void **state)
 
 	/* Column 0 (vpe index 0) is the smallest (total eof = 120 + 200) */
-	col = column_to_scan(segInfos, 4, numcols);
+	col = column_to_scan(segInfos, 4, numcols, InvalidRelation);
```
How about updating the unit-test and validate `AOCSDrop()` was called with the right arguments?

jimmyyih commented:
Updated unit test and squashed commits.  Still looking into the different in-memory states.

jimmyyih commented:
Squashing removed some important information that I had in refactor commits on top of the main 3 commits I'm trying to merge into master.  Noting it down here in a PR comment:

1. `column_to_scan()` dropping the segfiles in `AOSEG_STATE_AWAITING_DROP` state was okay but inserting tuples afterwards caused the tuples to go into segno 2 instead of the cleared segno 1. This is because the in-memory state was not correct. To fix this, we drop the hash entry when updating an aoco table.  There's a comment in `ATAocsNoRewrite()` for this.

2. `column_to_scan()` loops until it finds a segfile with nonzero tuplecount. It's very possible it could loop just once and the N-number of segfiles afterwards are just skipped (dangerously including the segfiles in `AOSEG_STATE_AWAITING_DROP` state). To fix this, we find the first segfile with nonzero tuplecount but continue the loop until the end to scan for segfiles in `AOSEG_STATE_AWAITING_DROP` state. Then the function returns a column as it previously would.

asimrp commented on src/backend/commands/tablecmds.c:
```diff
@@ -4734,6 +4754,23 @@ ATAocsNoRewrite(AlteredTableInfo *tab)
+		 * there is no rewrite because we may have dropped some
+		 * segfiles that were in AOSEG_STATE_AWAITING_DROP state. The
+		 * cost of recreating the entry later on is very low so this
+		 * should be fine.
```
Can we include a description of what will happen if the hash entry is not removed, so as to make the intent clear?

asimrp commented on src/backend/commands/test/tablecmds_test.c:
```diff
@@ -58,6 +72,14 @@ test__column_to_scan(void **state)
+	Gp_role = GP_ROLE_EXECUTE;
+	RelationData reldata;
+	expect_value(AOCSDrop, aorel, &reldata);
+	expect_check(AOCSDrop, compaction_segno, check_segno_list, drop_segno_list);
```
Nice!

jimmyyih commented:
Added more comments regarding to the introduced bugs I've fixed.  I have also added another somewhat related commit to this PR:
```
commit 7f5e915e5889eb84e3111640e13eb860839e0113
Author: Jimmy Yih <jyih@pivotal.io>
Date:   Tue May 9 13:29:03 2017 -0700

    Properly get smallest eof column during ADD COLUMN on AOCO table

    The column_to_scan function used to get the smallest eof column was
    operating incorrectly. It would loop the segfiles until it found a
    segfile with nonzero tuplecount, grab the first column that has a
    nonzero eof, and return that value. The issue here is that it did not
    iterate through all the columns to find the smallest nonzero eof. With
    this commit, we now loop through all the columns of the segfile to
    decide on a column to return. The unit tests for this function were
    also incorrect so we fix that in this commit too.
```

ashwinstar commented on src/backend/commands/tablecmds.c:
```diff
@@ -4604,11 +4604,10 @@ column_to_scan(AOCSFileSegInfo **segInfos, int nseg, int natts, Relation aocsrel
 			{
 				vpe = getAOCSVPEntry(segInfos[segi], i);
-				if (vpe->eof < min_eof && vpe->eof > 0)
+				if (vpe->eof > 0 && (!min_eof || vpe->eof < min_eof))
```
Aren't you using min_eof uninitialized here ?

jimmyyih replied:
Sure, fixed by initializing to 0 at the top of the function.

ashwinstar commented on src/backend/commands/test/tablecmds_test.c:
```diff
@@ -26,37 +26,39 @@ static int check_segno_list(const List* value, const List* check_value)
+	segInfos[0]->total_tupcount = 0;
 	segInfos[0]->vpinfo.nEntry = 3; /* number of columns */
-	segInfos[0]->vpinfo.entry[0].eof = 120;
+	segInfos[0]->vpinfo.entry[0].eof = 200;
```
This sounds little counter-initutive, is there are case where EOF > 0 and total_tupcount = 0 ? I doubt.

jimmyyih replied:
I think we need this here to make sure the flow actually skips segfiles with zero tuplecount.  If it does not skip in the function, the function will return column 2 instead of column 1 obtained in the next segfile.

ashwinstar replied:
Sure can keep it but I don't think that case can ever happen as if `total_tupcount = 0` EOF for all columns in it will be 0 as well, cannot be greater.

ashwinstar commented on src/backend/commands/tablecmds.c:
```diff
@@ -4604,11 +4604,10 @@ column_to_scan(AOCSFileSegInfo **segInfos, int nseg, int natts, Relation aocsrel
 			for (i = 0; i < natts; ++i)
```
Not really sure above if`segInfos[segi]->total_tupcount > 0` condition is needed ?

jimmyyih replied:
It should be there as the comment above it suggests:
```
## in utility mode on a primary segment (after delete and vacuum on AOCO table foo from master)
gptest=# select segno, column_num, tupcount, eof, eof_uncompressed, state from gp_toolkit.__gp_aocsseg_name('foo');
 segno | column_num | tupcount | eof | eof_uncompressed | state
-------+------------+----------+-----+------------------+-------
     1 |          0 |        0 |   0 |                0 |     1
     1 |          1 |        0 |   0 |                0 |     1
     2 |          0 |       12 |  88 |               88 |     1
     2 |          1 |       12 |  88 |               88 |     1
(4 rows)
```

In the code flow, it will skip segno 1 and go straight checking column eofs in segno 2.

ashwinstar replied:
Hmm okay its just for optimization to avoid looping `natts` times. Practically which can be achieved still using `eof > 0` itself by breaking if eof <= 0 move on to next segfile but fine.

jimmyyih commented:
@asimrp, it seems like we still need the states in question.  They control different things and operate under different phases of vacuum.  Reducing the number of states may cause issues for us down the road.

The three operate kinda like this...
```
COMPACTED_AWAITING_DROP (6):
 * 6 -> 5 start the drop phase to drop segfiles
 * 6 -> 8 in drop phase but skip because someone stopped me (no lock or serializable transactions)
 * 6 -> 7 in prepare phase, we found segfiles awaiting drop so lets try to drop them now

PSEUDO_COMPACTION_USE (7):
 * 7 -> 6 set to COMPACTED_AWAITING_DROP for drop phase to start dropping segfiles
 * 7 -> 2 transaction was aborted before drop phase started

COMPACTED_DROP_SKIPPED (8):
 * 8 -> 2 it's end of xact and we skipped drop phase... set to AWAITING_DROP state for next vacuum
```
PR #2242 should save us from this though.

jimmyyih commented:
Merged into master:
05f55e86d09800591f0cf6c26c9e6de1196d06c2
930b2c387c8a4a93cee8ca8d75b017e4b0717f37
cdc8172a835b1a7374ae5a94d426f6a34baf46ae
b1a1d1900814e646587f1081f90260ac2566388e
